client/widmgr/signer.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package widmgr
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type RPCer interface {
+	RPC(method string, args any, reply any) error
+}
+
+// IdentitySigner is the interface needed to retrieve signed identities for
+// workload identities. At runtime it is implemented by *widmgr.Signer.
+type IdentitySigner interface {
+	SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error)
+}
+
+// SignerConfig wraps the configuration parameters the workload identity manager
+// needs.
+type SignerConfig struct {
+	// NodeSecret is the node's secret token
+	NodeSecret string
+
+	// Region of the node
+	Region string
+
+	RPC RPCer
+}
+
+// Signer fetches and validates workload identities.
+type Signer struct {
+	nodeSecret string
+	region     string
+	rpc        RPCer
+}
+
+// NewSigner workload identity manager.
+func NewSigner(c SignerConfig) *Signer {
+	return &Signer{
+		nodeSecret: c.NodeSecret,
+		region:     c.Region,
+		rpc:        c.RPC,
+	}
+}
+
+// SignIdentities wraps the Alloc.SignIdentities RPC and retrieves signed
+// workload identities. The minIndex should be set to the lowest allocation
+// CreateIndex to ensure that the server handling the request isn't so stale
+// that it doesn't know the allocation exist (and therefore rejects the signing
+// requests).
+//
+// Since a single rejection causes an error to be returned, SignIdentities
+// should currently only be used when requesting signed identities for a single
+// allocation.
+func (s *Signer) SignIdentities(minIndex uint64, req []*structs.WorkloadIdentityRequest) ([]*structs.SignedWorkloadIdentity, error) {
+	args := structs.AllocIdentitiesRequest{
+		Identities: req,
+		QueryOptions: structs.QueryOptions{
+			Region:        s.region,
+			MinQueryIndex: minIndex - 1,
```

Comment:
Why -1?

Reply:
I honestly don't know, I lifted this from Schmichael's old code. @schmichael can you shed some light on this?

Reply:
I should have left a comment because this took me way too long to remember. :grimacing: 

Normally blocking queries are used to wait for changes *subsequent* to the `MinQueryIndex` (the `?index=...` in HTTP APIs). So if an Allocation was created at index 123, you would want to block with index=123 to ensure the query returned only when the allocation was modified *after* creation.

This isn't our normal blocking query though. I think it's the first time we've taken a different approach:

Nomad determines the validity of a workload identity based on the liveness of the workload (allocation) it represents. Therefore Clients requesting signed workload identities via the `Alloc.SignIdentities` RPC want to ensure the Server doesn't erroneously reject the request because the Server's raft index is out of stale (either due to the server being mid-restore or just due to "fun" distributed systems race conditions).

Therefore the Client sets `MinQueryIndex = alloc.CreateIndex - 1` to instruct the Server to block at least _until the Allocation is created._ The Client doesn't care about modifications after that (or more accurately: modifications after that, like `DesiredStatus=stop`, will be handled by *other* RPCs the Client makes).

So this RPC, unlike all other RPCs AFAIK, doesn't care about "subsequent modifications" after an index. It's just trying to ensure the Server's state isn't too stale to know about the alloc.

...now to figure out how to describe this succinctly in a comment. :face_with_spiral_eyes:

Reply:
Thanks for clarifying @schmichael, I summarized in a comment in [e8ef776](https://github.com/hashicorp/nomad/pull/18431/commits/e8ef7763782f937b3915c163134604544919f735), I hope it's clear enough.